### PR TITLE
Add option to build ROSCO as an external project

### LINKED ISSUE #261: Get ROSCO working in OpenTurbine
Update Controller interface to work with ROSCO and understand which inputs need to be supplied with it specifically. The controller input file is available at https://github.com/IEAWindTask37/IEA-15-240-RWT/blob/master/OpenFAST/IEA-15-240-RWT-Monopile/IEA-15-240-RWT-Monopile_DISCON.IN. This could be included as an external project via CMake.

### PR DESCRIPTION
**Pull request by deslaughter**
This PR builds on the ADI library PR to allow building of ROSCO as an external project. This functionality will make it easier to deploy a version of OpenTurbine which can run a complete turbine simulation with minimal input from the user. Completion of this PR will close #261 .

### CODE REVIEW

File: cmake/OpenTurbineOptions.cmake [hunk 1]
```diff
@@ -16,7 +16,8 @@ macro(openturbine_setup_options)
   option(OpenTurbine_ENABLE_CPPCHECK "Enable CppCheck analysis" OFF)
   option(OpenTurbine_ENABLE_PCH "Enable precompiled headers" OFF)
   option(OpenTurbine_ENABLE_VTK "Use VTK for visualization" OFF)
-  option(OpenTurbine_BUILD_OPENFAST_ADI "Build the OpenFAST ADI external project" ON)
+  option(OpenTurbine_BUILD_OPENFAST_ADI "Build the OpenFAST ADI external project" OFF)
+  option(OpenTurbine_BUILD_ROSCO_CONTROLLER "Build the ROSCO controller external project" ON)
```
Comment (ddement) [inline cmake/OpenTurbineOptions.cmake:20]: I'm thinking anything that builds an external project should be off by default.  What do you think?

Comment (deslaughter) [reply cmake/OpenTurbineOptions.cmake:20]: I agree. I'll change it to OFF by default

File: tests/unit_tests/CMakeLists.txt [hunk 2]
```diff
@@ -7,6 +7,7 @@ add_subdirectory(model)
 add_subdirectory(regression)
 add_subdirectory(solver)
 add_subdirectory(system)
+add_subdirectory(external)
```
Comment (faisal-bhuiyan) [inline tests/unit_tests/CMakeLists.txt:10]: Let's keep things alphabetic there.

File: cmake/OpenTurbineOptions.cmake [hunk 3]
```diff
@@ -16,7 +16,8 @@ macro(openturbine_setup_options)
   option(OpenTurbine_ENABLE_CPPCHECK "Enable CppCheck analysis" OFF)
   option(OpenTurbine_ENABLE_PCH "Enable precompiled headers" OFF)
   option(OpenTurbine_ENABLE_VTK "Use VTK for visualization" OFF)
-  option(OpenTurbine_BUILD_OPENFAST_ADI "Build the OpenFAST ADI external project" ON)
+  option(OpenTurbine_BUILD_OPENFAST_ADI "Build the OpenFAST ADI external project" OFF)
```
Comment (faisal-bhuiyan) [inline cmake/OpenTurbineOptions.cmake:19]: Let's add this flag to the CI please.

File: tests/unit_tests/external/test_rosco_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+
+namespace openturbine::tests {
+
+TEST(ROSCO_Controller, initialize) {
+}
```
Comment (faisal-bhuiyan) [inline tests/unit_tests/external/test_rosco_controller.cpp:7]: Are you planning to add anything here for this PR?

**Review by faisal-bhuiyan — APPROVED** [review]

File: tests/unit_tests/external/test_rosco_controller.cpp (hunk 4 above)
Comment (deslaughter) [reply tests/unit_tests/external/test_rosco_controller.cpp:23]: Yes, added in more recent commit

File: src/constraints/calculate_constraint_output.hpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "constraint_type.hpp"
+
+#include "src/math/quaternion_operations.hpp"
+#include "src/math/vector_operations.hpp"
+#include "src/types.hpp"
+
+namespace openturbine {
+
+struct CalculateConstraintOutput {
```
Comment (ddement) [inline src/constraints/calculate_constraint_output.hpp:13]: Is there a better name here than "output"?  Perhaps `CalculateConstraintAngularMotion` or something along those lines?

**Review by ddement — APPROVED** [review]

File: tests/unit_tests/regression/test_math.cpp [hunk 6]
```diff
@@ -203,6 +203,39 @@ TEST(QuaternionTest, RotationVectorToQuaternion_Set2) {
...
+    const auto dtheta = M_PI / static_cast<double>(n);
+    for (size_t i = 0; i < n; ++i) {
+        Array_3 rot_vec{static_cast<double>(i) * dtheta, 0., 0.};
+        auto q = RotationVectorToQuaternion(rot_vec);
+        auto rot_vec2 = QuaternionToRotationVector(q);
+        ASSERT_NEAR(rot_vec2[0], rot_vec[0], 1e-7);
+        ASSERT_NEAR(rot_vec2[1], rot_vec[1], 1e-7);
+        ASSERT_NEAR(rot_vec2[2], rot_vec[2], 1e-7);
+    }
+}
```
Comment (ddement) [inline tests/unit_tests/regression/test_math.cpp:237]: another way to have simplified this for CI and maintain your old test coverage would be to make the body of the for loop its own function:
```
test_quaternion_to_rotation_vector_2(size_t i, size_t j, size_t k) {
...
}
TEST{QuaternionTest, QuaternionToRotationVector_2) {
    const auto n = 100U;
    const auto dtheta = M_PI / static_cast<double>(n);
    for(size_t i = 0; i < n; ++i) {
        for(size_t j = 0; j < n; ++j) {
            for(size_t k = 0; k < n; ++k) {
                test_quaternion_to_rotation_vector_2(i, j, k);
            }
        }
    }
}
```

File: src/constraints/calculate_constraint_output.hpp (hunk 5 above)
Comment (deslaughter) [reply src/constraints/calculate_constraint_output.hpp:13]: Possibly, but I could see a possibility of having output for other constraint types, so I am leaving it general until I rule that out.

File: tests/unit_tests/regression/test_math.cpp (hunk 6 above)
Comment (deslaughter) [reply tests/unit_tests/regression/test_math.cpp:237]: I'll give that a try

File: src/constraints/calculate_constraint_output.hpp (hunk 5 above)
Comment (ddement) [reply src/constraints/calculate_constraint_output.hpp:13]: Sounds good.

**Pull request merged**